crates/tarball/src/lib.rs

```diff
@@ -201,73 +201,88 @@ impl<'a> DownloadTarballToStore<'a> {
-                entry.read_to_end(&mut buffer).unwrap();
-
-                let entry_path = entry.path().unwrap();
-                let cleaned_entry_path =
-                    entry_path.components().skip(1).collect::<PathBuf>().into_os_string();
-                let (file_path, file_hash) = store_dir
-                    .write_cas_file(&buffer, file_is_executable)
-                    .map_err(TarballError::WriteCasFile)?;
-
-                let tarball_index_key = cleaned_entry_path
-                    .to_str()
-                    .expect("entry path must be valid UTF-8") // TODO: propagate this error, provide more information
-                    .to_string(); // TODO: convert cleaned_entry_path to String too.
-
-                if let Some(previous) = cas_paths.insert(cleaned_entry_path, file_path) {
-                    tracing::warn!(?previous, "Duplication detected. Old entry has been ejected");
+            rayon::scope(|scope| {
+                verify_checksum(&response, package_integrity.clone())
+                    .map_err(TaskError::Checksum)?;
+
+                // TODO: move tarball extraction to its own function
+                // TODO: test it
+                // TODO: test the duplication of entries
+
+                let mut archive = decompress_gzip(&response, package_unpacked_size)
+                    .map_err(TaskError::Other)?
+                    .pipe(Cursor::new)
+                    .pipe(Archive::new);
+
+                let entries = archive
+                    .entries()
+                    .map_err(TarballError::ReadTarballEntries)
+                    .map_err(TaskError::Other)?
+                    .filter(|entry| !entry.as_ref().unwrap().header().entry_type().is_dir());
+
+                let ((_, Some(capacity)) | (capacity, None)) = entries.size_hint();
+                let mut cas_paths = HashMap::<OsString, PathBuf>::with_capacity(capacity);
+                let mut pkg_files_idx =
+                    PackageFilesIndex { files: HashMap::with_capacity(capacity) };
+
+                for entry in entries {
+                    let mut entry = entry.unwrap();
+
+                    let file_mode = entry.header().mode().expect("get mode"); // TODO: properly propagate this error
+                    let file_is_executable = file_mode::is_all_exec(file_mode);
+
+                    // Read the contents of the entry
+                    let mut buffer = Vec::with_capacity(entry.size() as usize);
+                    entry.read_to_end(&mut buffer).unwrap();
+
+                    let entry_path = entry.path().unwrap();
+                    let cleaned_entry_path =
+                        entry_path.components().skip(1).collect::<PathBuf>().into_os_string();
+                    let (file_path, file_hash) = store_dir
+                        .write_cas_file(scope, buffer, file_is_executable)
+                        .map_err(TarballError::WriteCasFile)?;
```

Comment:
This seems to be the only IO operation in the entire scope. And none of the operations that follow depend on this IO operation being done either. Meaning, we could potentially defer this operation after building the `cas_paths` (i.e. do the IO on the created `cas_paths`). See if this suggestion can improve the code quality and performance.